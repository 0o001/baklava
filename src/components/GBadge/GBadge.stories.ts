import { storiesOf } from '@storybook/vue';
import GBadge from './GBadge.vue';
import GIcon from '../GIcon/GIcon.vue';
import { withKnobs, boolean, text } from '@storybook/addon-knobs';

const stories = storiesOf('GBadge', module);
stories.addDecorator(withKnobs);
stories
  .addParameters({ component: GBadge })
  .add('Default', () => ({
    props: {
      border: {
        default: boolean('Border', false),
      },
      inline: {
        default: boolean('Inline', false),
      },
      dot: {
        default: boolean('Dot', false),
      },
      left: {
        default: boolean('Left', false),
      },
      bottom: {
        default: boolean('Bottom', false),
      },
      overlap: {
        default: boolean('Overlap', false),
      },
      backgroundColor: {
        default: text('Background Color', 'orange-100'),
      },
      borderColor: {
        default: text('Border Color', 'mid-grey-100'),
      },
      fontColor: {
        default: text('Font Color', 'white'),
      },
      text: {
        default: text('Text', '2'),
      },
      showBadge: {
        default: boolean('ShowBadge', true),
      },
    },
    components: { GBadge, GIcon },
    template: `
      <div style='padding:30px; display: flex; justify-content: space-between;'>
        <GBadge
          :text="text"
          :border="border"
          :left="left"
          :bottom="bottom"
          :overlap="overlap"
          :dot="dot"
          :inline="inline"
          :background-color="backgroundColor"
          :font-color="fontColor"
          :border-color="borderColor"
          :show-badge="showBadge"
        >
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge text="2" :left="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge text="2" :left="true" :bottom="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge text="2" :bottom="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge text="2" :overlap="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge text="2" :left="true" :overlap="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge text="2" :left="true" :bottom="true" :overlap="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge text="2" :bottom="true" :overlap="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
      </div>`,
  }))
  .add('Dot', () => ({
    components: { GBadge, GIcon },
    template: `
      <div style='padding:30px; display: flex; justify-content: space-between;'>
        <GBadge :dot="true" background-color="orange-100">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge :dot="true" :left="true" background-color="orange-100">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge :dot="true" :left="true" :bottom="true" background-color="orange-100">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge :dot="true" :bottom="true" background-color="orange-100">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge :dot="true" background-color="orange-100" :overlap="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge :dot="true" :left="true" background-color="orange-100" :overlap="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge :dot="true" :left="true" :bottom="true" background-color="orange-100" :overlap="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
        <GBadge :dot="true" :bottom="true" background-color="orange-100" :overlap="true">
          <template>
            <GIcon name="bell"/>
          </template>
        </GBadge>
      </div>`,
  }));
