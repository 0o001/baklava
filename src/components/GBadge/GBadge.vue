<template>
  <div
    :class="badgeContainerClass"
  >
    <div
      v-if="dot"
      :class="badgeClass"
    />
    <div
      v-else
      :class="badgeClass"
    >
      <GText
        v-if="text"
        variant="caption"
        :color="textColor"
      >
        {{ text }}
      </GText>
    </div>
    <div class="content">
      <slot />
    </div>
  </div>
</template>

<script>

import GText from '../GText/GText.vue';
export default {
  name: 'GBadge',
  components: { GText },
  props: {
    border: {
      type: Boolean,
      default: false,
    },
    inline: {
      type: Boolean,
      default: false,
    },
    dot: {
      type: Boolean,
      default: false,
    },
    left: {
      type: Boolean,
      default: false,
    },
    bottom: {
      type: Boolean,
      default: false,
    },
    overlap: {
      type: Boolean,
      default: false,
    },
    backgroundColor: {
      type: String,
      default: 'orange-100',
    },
    borderColor: {
      type: String,
      default: 'mid-grey-100',
    },
    textColor: {
      type: String,
      default: 'white',
    },
    text: {
      type: String,
      default: '0',
    },
  },
  computed: {
    badgeClass () {
      return {
        badge: true,
        overlap: this.overlap,
        bottom: this.bottom,
        left: this.left,
        dot: this.dot,
        inline: this.inline,
        border: this.border && !this.dot,
        [`g-border-${this.borderColor}`]: true,
        [`g-bg-${this.backgroundColor}`]: true,
      };
    },
    badgeContainerClass () {
      return {
        'g-badge': true,
        inline: this.inline,
        left: this.left,
      };
    },
  },
};
</script>

<style lang="scss" scoped>
  .g-badge{
    position: relative;
    display: inline-block;
    line-height: 1;
    &.inline{
      display: inline-flex;
      align-items: center;
      flex-direction: row-reverse;
      &.left{
        flex-direction: row;
      }
    }
    & .badge{
      position: absolute;
      text-align: center;
      min-width: 20px;
      height: 20px;
      padding: 3px 0;
      border-radius: 10px;
      bottom: calc(100% - 6px);
      left: calc(100% - 6px);
      right: auto;
      top: auto;

      &.left{
        right: calc(100% - 6px);
        left: auto;
      }
      &.bottom{
        top: calc(100% - 6px);
        bottom: auto;
      }
      &.border{
        border: 2px solid;
        padding: 1px 0;
      }
      &.overlap {
        bottom: calc(100% - 10px);
        left: calc(100% - 10px);
        right: auto;
        top: auto;
        &.left{
          right: calc(100% - 10px);;
          left: auto;
        }
        &.bottom{
          top: calc(100% - 10px);;
          bottom: auto;
        }
      }
      &.dot{
        min-width: 8px;
        width: 8px;
        height: 8px;
        padding: 0;
        &:not(.overlap){
          top: -6px;
          right: -6px;
          left: auto;
          bottom: auto;
          &.left{
            right: auto;
            left: -6px;
          }
          &.bottom{
            top: auto;
            bottom: -6px;
          }
        }
        &.overlap{
          top: -2px;
          right: -2px;
          left: auto;
          bottom: auto;
          &.left{
            right: auto;
            left: -2px;
          }
          &.bottom{
            top: auto;
            bottom: -2px;
          }
        }
      }
      &.inline{
        position: relative;
        margin: 0 4px;
        top: auto;
        left: auto;
        right: auto;
        bottom: auto;

      }
    }
  }
</style>
