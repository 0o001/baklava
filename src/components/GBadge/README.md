# GBadge

## Slots

| Name      | Description |
| --------- | ----------- |
| `default` | &nbsp;      |

## Props

| Name               | Type      | Description | Default          |
| ------------------ | --------- | ----------- | ---------------- |
| `border`           | `Boolean` |             | `false`          |
| `inline`           | `Boolean` |             | `false`          |
| `dot`              | `Boolean` |             | `false`          |
| `left`             | `Boolean` |             | `false`          |
| `bottom`           | `Boolean` |             | `false`          |
| `overlap`          | `Boolean` |             | `false`          |
| `background-color` | `String`  |             | `"orange-100"`   |
| `border-color`     | `String`  |             | `"mid-grey-100"` |
| `text-color`       | `String`  |             | `"white"`        |
| `text`             | `String`  |             | `"0"`            |

## Computed Properties

| Name                  | Type     | Description                                                                                                |
| --------------------- | -------- | ---------------------------------------------------------------------------------------------------------- |
| `badgeClass`          | `object` | **Dependencies:** `overlap`, `bottom`, `left`, `dot`, `inline`, `border`, `borderColor`, `backgroundColor` |
| `badgeContainerClass` | `object` | **Dependencies:** `inline`, `left`                                                                         |

