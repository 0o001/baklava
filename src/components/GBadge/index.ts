import GBadge from './GBadge.vue';

export default GBadge;
