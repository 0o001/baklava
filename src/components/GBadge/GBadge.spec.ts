import { shallowMount } from '@vue/test-utils';
import GBadge from './';

describe('GBadge', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallowMount(GBadge);
  });

  it('should match snapshot with default style', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render with default styles', () => {
    expect(wrapper.vm.badgeClass).toEqual({
      badge: true,
      border: false,
      bottom: false,
      dot: false,
      'g-bg-orange-100': true,
      inline: false,
      left: false,
      overlap: false,
    });
    expect(wrapper.vm.badgeContainerClass).toEqual({
      'g-badge': true,
      inline: false,
      left: false,
    });
  });

  it('should render with given background and border color', async () => {
    // given
    wrapper.setProps({
      backgroundColor: 'green-400',
      borderColor: 'green-500',
    });

    // when
    await wrapper.vm.$nextTick();

    // then
    expect(wrapper.vm.badgeClass).toEqual({
      badge: true,
      border: false,
      bottom: false,
      dot: false,
      'g-bg-green-400': true,
      inline: false,
      left: false,
      overlap: false,
    });
  });

  it('should render with given inline and left', async () => {
    // given
    wrapper.setProps({
      inline: true,
      left: true,
    });

    // when
    await wrapper.vm.$nextTick();

    // then
    expect(wrapper.vm.badgeContainerClass).toEqual({
      'g-badge': true,
      inline: true,
      left: true,
    });
  });
});
