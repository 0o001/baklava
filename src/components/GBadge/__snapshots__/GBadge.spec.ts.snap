// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GBadge should match snapshot with default style 1`] = `
<div
  class="g-badge"
>
  <div
    class="badge g-border-mid-grey-100 g-bg-orange-100"
  >
    <gtext-stub
      color="white"
      variant="caption"
    >
      
      0
    
    </gtext-stub>
  </div>
   
  <div
    class="content"
  />
</div>
`;
